zkp/golang/internal/smt/merkletree.go

```diff
@@ -174,14 +194,15 @@ func (mt *sparseMerkleTree) addLeaf(newLeaf core.Node, currentNodeIndex core.Nod
 	}
 
 	var nextKey core.NodeIndex
-	currentNode, err := mt.GetNode(currentNodeIndex)
+	currentNode, err := mt.getNode(currentNodeIndex)
 	if err != nil {
 		return nil, err
 	}
 	switch currentNode.Type() {
 	case core.NodeTypeEmpty:
-		// We have searched to the bottom level and are ensured that
-		// the node doesn't exist yet. We can add the new leaf node
+		// We have searched to a level and have found a position in the
+		// index's path where the tree is empty. This means we have found
+		// the node that doesn't exist yet. We can add the new leaf node here
 		return mt.addNode(newLeaf)
```

Comment:
Do you need a write lock?

Reply:
It's confusing because inside `addNode` you also do a `getNode` and none of these have a lock so you are always relying on the calling function to lock which is dangerous. I understand why you are doing it because in the`addLeaf` function you need to iterate and you don't want to lock and unlock on each iteration.

Reply:
the nested `addLeaf()` calls are the main reason for now doing more granular lock/unlock. The series of updates from the lead node up to the root with the branch nodes along the way must be atomic, otherwise we end up with a broken/inconsistent tree. thus I felt having an overall write lock on the calling function is the best way to go.